taichi/analysis/count_statements.cpp

```diff
@@ -0,0 +1,63 @@
+#include "taichi/ir/ir.h"
+
+TLANG_NAMESPACE_BEGIN
+
+// Count all statements (including containers)
+class StmtCounter : public BasicStmtVisitor {
+ private:
+  StmtCounter() {
+    counter = 0;
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  using BasicStmtVisitor::visit;
+
+ public:
+  void visit(Stmt *stmt) override {
+    counter++;
+  }
+
+  void visit(IfStmt *stmt) override {
+    counter++;
+    BasicStmtVisitor::visit(stmt);
```

Comment:
I didn't realize we can call this when writing visitors. In this way, many container stmts' visitor function can be
```C++
do_something(stmt);
BasicStmtVisitor::visit(stmt);
```

And we can write a single function `void visit(ContainerStmt *stmt)` for all container stmts if the container statements are inherited from something like `class ContainerStmt: public Stmt`.

Reply:
Or we can simply use `Stmt::is_container_statement()` to make the inheritance hierarchy simple?

Reply:
Oh yes. I'll open a PR to make `verify` look better soon.

Reply:
Wait... Is this correct?
```C++
void visit(Stmt *stmt) override {
  basic_verify(stmt);
  if (stmt->is_container_statement()) {
    BasicStmtVisitor::visit(stmt);
  }
}
```
Would `IfStmt::accept` call `BasicStmtVisitor::visit(IfStmt*)`? And can `BasicStmtVisitor::visit()` deduce the actual type?

Reply:
`IfStmt::accept` indeed calls `BasicStmtVisitor::visit(IfStmt*)`...

Reply:
Oh sorry - I was only looking at your code and thought you were using `BasicStmtVisitor::visit(stmt)`. Yes, if you do use `accept`, it will invoke the right visitor.

Reply:
What I want to do is to make the visitor function of all container stmts call `basic_verify(stmt);` first, and then call `BasicStmtVisitor::visit(stmt)`.

However, if I don't override `BasicStmtVisitor::visit(IfStmt*)`, `IfStmt::accept` cannot call `IRVerifier::visit(Stmt*)`... So I can't remove all the visitor methods in `IRVerifier` as a simplification of the code.